In [1]:
DATASET_PATH = '../data/freelancer_earnings_bd.csv'

In [2]:
import pandas as pd

df = pd.read_csv(DATASET_PATH)

In [3]:
df

Freelancer_ID       Job_Category       Platform Experience_Level  \
0                 1    Web Development         Fiverr         Beginner   
1                 2    App Development         Fiverr         Beginner   
2                 3    Web Development         Fiverr         Beginner   
3                 4         Data Entry  PeoplePerHour     Intermediate   
4                 5  Digital Marketing         Upwork           Expert   
...             ...                ...            ...              ...   
1945           1946     Graphic Design         Fiverr         Beginner   
1946           1947                SEO         Upwork     Intermediate   
1947           1948                SEO  PeoplePerHour           Expert   
1948           1949                SEO     Freelancer     Intermediate   
1949           1950     Graphic Design     Freelancer           Expert   

     Client_Region  Payment_Method  Job_Completed  Earnings_USD  Hourly_Rate  \
0             Asia  Mobile Banking            180          1620        95.79   
1        Australia  Mobile Banking            218          9078        86.38   
2               UK          Crypto             27          3455        85.17   
3             Asia   Bank Transfer             17          5577        14.37   
4             Asia          Crypto            245          5898        99.37   
...            ...             ...            ...           ...          ...   
1945           USA  Mobile Banking            143          6823        22.54   
1946   Middle East          Crypto            164          7942        77.20   
1947            UK   Bank Transfer            236          9838        24.64   
1948        Europe   Bank Transfer            152          4492        71.07   
1949   Middle East          PayPal            247          9297        22.22   

      Job_Success_Rate  Client_Rating  Job_Duration_Days Project_Type  \
0                68.73           3.18                  1        Fixed   
1                97.54           3.44                 54        Fixed   
2                86.60           4.20                 46       Hourly   
3                79.93           4.47                 41       Hourly   
4                57.80           5.00                 41       Hourly   
...                ...            ...                ...          ...   
1945             75.86           4.65                 13       Hourly   
1946             72.01           3.29                 34       Hourly   
1947             57.37           4.67                 38        Fixed   
1948             66.41           4.08                 70        Fixed   
1949             71.70           4.69                 68        Fixed   

      Rehire_Rate  Marketing_Spend  
0           40.19               53  
1           36.53              486  
2           74.05              489  
3           27.58               67  
4           69.09              489  
...           ...              ...  
1945        26.55              133  
1946        24.81              343  
1947        61.52              370  
1948        32.40               32  
1949        45.42              236  

[1950 rows x 15 columns]

1. Насколько выше доход у фрилансеров, принимающих оплату в криптовалюте, по сравнению с другими способами оплаты?

In [4]:
income_by_payment = df.groupby('Payment_Method')['Earnings_USD'].agg([
            'mean', 'median', 'count', 'std'
]).round(2)
income_by_payment

mean  median  count      std
Payment_Method                                 
Bank Transfer   5019.96  4903.0    494  2962.84
Crypto          5139.30  5224.0    514  2926.06
Mobile Banking  4923.65  4965.5    476  2972.94
PayPal          4976.69  4982.0    466  2843.11

In [5]:
income_by_payment.to_dict()

{'mean': {'Bank Transfer': 5019.96,
  'Crypto': 5139.3,
  'Mobile Banking': 4923.65,
  'PayPal': 4976.69},
 'median': {'Bank Transfer': 4903.0,
  'Crypto': 5224.0,
  'Mobile Banking': 4965.5,
  'PayPal': 4982.0},
 'count': {'Bank Transfer': 494,
  'Crypto': 514,
  'Mobile Banking': 476,
  'PayPal': 466},
 'std': {'Bank Transfer': 2962.84,
  'Crypto': 2926.06,
  'Mobile Banking': 2972.94,
  'PayPal': 2843.11}}

2. Как распределяется доход фрилансеров в зависимости от региона проживания?

In [6]:
# 2. Распределение дохода по регионам
df.groupby('Client_Region')['Earnings_USD'].agg([
    'mean', 'median', 'count', 'std'
]).round(2)

mean  median  count      std
Client_Region                                 
Asia           5172.28  5297.0    281  3032.98
Australia      4966.10  5296.0    298  2908.80
Canada         5350.13  5435.5    246  2883.31
Europe         4890.53  5000.0    262  3050.31
Middle East    4870.82  4971.0    291  2849.23
UK             5047.09  4872.5    280  2923.10
USA            4872.95  4643.5    292  2839.25

In [7]:
df.groupby('Client_Region')['Earnings_USD'].agg([
    'mean', 'median', 'count', 'std'
]).round(2)

mean  median  count      std
Client_Region                                 
Asia           5172.28  5297.0    281  3032.98
Australia      4966.10  5296.0    298  2908.80
Canada         5350.13  5435.5    246  2883.31
Europe         4890.53  5000.0    262  3050.31
Middle East    4870.82  4971.0    291  2849.23
UK             5047.09  4872.5    280  2923.10
USA            4872.95  4643.5    292  2839.25

In [8]:
result_dict = (df.groupby('Client_Region')['Earnings_USD']
                .agg(['mean', 'count', 'std'])
                .round(2)
                .reset_index()
                .to_dict('list'))

final_dict = {
    'Client_Region': result_dict['Client_Region'],
    'Earnings_USD_mean': result_dict['mean'],
    'Earnings_USD_count': result_dict['count'],
    'Earnings_USD_std': result_dict['std']
}
final_dict

{'Client_Region': ['Asia',
  'Australia',
  'Canada',
  'Europe',
  'Middle East',
  'UK',
  'USA'],
 'Earnings_USD_mean': [5172.28,
  4966.1,
  5350.13,
  4890.53,
  4870.82,
  5047.09,
  4872.95],
 'Earnings_USD_count': [281, 298, 246, 262, 291, 280, 292],
 'Earnings_USD_std': [3032.98,
  2908.8,
  2883.31,
  3050.31,
  2849.23,
  2923.1,
  2839.25]}

3. Какой процент фрилансеров, считающих себя экспертами, выполнил менее 100 проектов?

In [9]:
experts = df[(df['Experience_Level'] == 'Expert')]
# experts = experts[['Experience_Level', 'Job_Completed']]
len(experts)

641

In [10]:
from typing import Dict, List
def get_freelancers_where(condition: str, page_content: str, fields: List = []):
    relevant_part: Dict = df.query(condition)
    if fields:
        relevant_part = relevant_part[fields]
    return {
        'page_content': f'{page_content}: {str(relevant_part.to_dict())}',
    }

get_freelancers_where(
    'Experience_Level == "Expert"', 
    page_content='Фрилансеры, считающие себя экспертами и их количество проектов (id, Job_Completed)',
    fields=['Job_Completed'],
)

{'page_content': "Фрилансеры, считающие себя экспертами и их количество проектов (id, Job_Completed): {'Job_Completed': {4: 245, 16: 18, 18: 75, 21: 110, 24: 149, 26: 178, 31: 33, 32: 279, 33: 289, 34: 74, 35: 283, 39: 184, 40: 268, 43: 253, 47: 116, 49: 40, 52: 187, 53: 89, 54: 273, 56: 94, 58: 26, 60: 219, 65: 158, 70: 157, 73: 149, 77: 67, 79: 19, 81: 125, 82: 132, 83: 292, 84: 184, 85: 86, 87: 28, 88: 172, 90: 74, 94: 221, 96: 145, 98: 36, 99: 156, 104: 102, 106: 94, 111: 285, 112: 283, 113: 240, 114: 132, 115: 213, 117: 173, 131: 141, 132: 277, 134: 192, 135: 145, 136: 135, 139: 30, 142: 233, 145: 176, 147: 299, 150: 218, 152: 208, 156: 66, 157: 62, 158: 278, 159: 112, 160: 191, 162: 52, 166: 258, 174: 39, 175: 221, 181: 94, 183: 163, 185: 286, 195: 108, 200: 167, 203: 72, 204: 250, 205: 142, 206: 291, 211: 72, 212: 191, 221: 53, 223: 99, 224: 29, 226: 183, 232: 257, 233: 76, 236: 287, 238: 250, 242: 192, 246: 129, 248: 17, 251: 267, 252: 10, 254: 34, 256: 160, 261: 243, 270: 158,

In [11]:
# optional
round((len(experts[experts['Job_Completed'] < 100]) / len(experts)) * 100, 2)

33.85

4. Какая корреляция между опытом работы и доходом?

In [12]:
df[['Job_Duration_Days', 'Earnings_USD']].corr().iloc[0, 1].round(2)

np.float64(0.01)

In [13]:
df.groupby('Job_Duration_Days')['Earnings_USD'].agg(['mean', 'count', 'std']).round(2)

mean  count      std
Job_Duration_Days                         
1                  4235.13     31  2720.46
2                  6298.88     25  2518.29
3                  5130.96     23  2911.92
4                  5257.61     18  2296.43
5                  4920.43     23  3264.68
...                    ...    ...      ...
85                 4438.39     28  2811.92
86                 4549.07     15  3039.82
87                 5134.07     15  3015.54
88                 5292.27     30  3004.52
89                 4439.69     29  2766.02

[89 rows x 3 columns]

In [14]:
result_dict = (df.groupby('Job_Duration_Days')['Earnings_USD']
                .agg(['mean', 'count', 'std'])
                .round(2)
                .reset_index()
                .to_dict('list'))

In [15]:
final_dict = {
    'Job_Duration_Days': result_dict['Job_Duration_Days'],
    'Earnings_USD_mean': result_dict['mean'],
    'Earnings_USD_count': result_dict['count'],
    'Earnings_USD_std': result_dict['std']
}
final_dict

{'Job_Duration_Days': [1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
  25,
  26,
  27,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  38,
  39,
  40,
  41,
  42,
  43,
  44,
  45,
  46,
  47,
  48,
  49,
  50,
  51,
  52,
  53,
  54,
  55,
  56,
  57,
  58,
  59,
  60,
  61,
  62,
  63,
  64,
  65,
  66,
  67,
  68,
  69,
  70,
  71,
  72,
  73,
  74,
  75,
  76,
  77,
  78,
  79,
  80,
  81,
  82,
  83,
  84,
  85,
  86,
  87,
  88,
  89],
 'Earnings_USD_mean': [4235.13,
  6298.88,
  5130.96,
  5257.61,
  4920.43,
  5142.6,
  4927.58,
  4770.22,
  4913.0,
  4782.68,
  4585.92,
  4594.67,
  5261.72,
  4292.64,
  4980.31,
  4847.66,
  5062.5,
  4541.17,
  5239.2,
  4344.57,
  4743.65,
  4939.45,
  6587.76,
  4979.18,
  5522.57,
  3571.33,
  5556.04,
  4658.75,
  4167.0,
  5787.21,
  5221.29,
  5476.94,
  3621.63,
  5025.58,
  4786.79,
  6234.9,
  4344.91,
  5516.58,
  5480.4,
  4701.0,


[Document(id='2', metadata={}, page_content="Фрилансеры, считающие себя экспертами и их количество проектов (id, Job_Completed): {'Job_Completed': {4: 245, 16: 18, 18: 75, 21: 110, 24: 149, 26: 178, 31: 33, 32: 279, 33: 289, 34: 74, 35: 283, 39: 184, 40: 268, 43: 253, 47: 116, 49: 40, 52: 187, 53: 89, 54: 273, 56: 94, 58: 26, 60: 219, 65: 158, 70: 157, 73: 149, 77: 67, 79: 19, 81: 125, 82: 132, 83: 292, 84: 184, 85: 86, 87: 28, 88: 172, 90: 74, 94: 221, 96: 145, 98: 36, 99: 156, 104: 102, 106: 94, 111: 285, 112: 283, 113: 240, 114: 132, 115: 213, 117: 173, 131: 141, 132: 277, 134: 192, 135: 145, 136: 135, 139: 30, 142: 233, 145: 176, 147: 299, 150: 218, 152: 208, 156: 66, 157: 62, 158: 278, 159: 112, 160: 191, 162: 52, 166: 258, 174: 39, 175: 221, 181: 94, 183: 163, 185: 286, 195: 108, 200: 167, 203: 72, 204: 250, 205: 142, 206: 291, 211: 72, 212: 191, 221: 53, 223: 99, 224: 29, 226: 183, 232: 257, 233: 76, 236: 287, 238: 250, 242: 192, 246: 129, 248: 17, 251: 267, 252: 10, 254: 34, 256: 160, 261: 243, 270: 158, 271: 278, 273: 141, 275: 239, 277: 111, 280: 298, 284: 207, 287: 18, 289: 199, 290: 118, 291: 163, 295: 41, 299: 71, 300: 34, 301: 100, 306: 36, 307: 43, 312: 78, 315: 142, 319: 30, 326: 214, 327: 119, 329: 269, 332: 272, 333: 120, 334: 92, 339: 277, 340: 123, 341: 28, 342: 262, 343: 96, 345: 153, 347: 57, 349: 269, 350: 266, 357: 40, 358: 227, 361: 192, 363: 146, 368: 133, 369: 261, 374: 43, 375: 267, 381: 223, 386: 25, 388: 229, 393: 156, 395: 194, 396: 26, 402: 229, 403: 127, 405: 127, 413: 135, 414: 73, 415: 254, 416: 181, 418: 21, 419: 164, 420: 74, 421: 117, 423: 25, 432: 198, 436: 293, 437: 135, 444: 250, 449: 60, 450: 52, 452: 63, 453: 81, 455: 59, 458: 89, 461: 297, 462: 53, 463: 71, 466: 33, 469: 184, 470: 155, 472: 189, 473: 138, 474: 181, 475: 26, 477: 81, 479: 8, 481: 132, 484: 239, 485: 140, 486: 172, 490: 160, 493: 251, 495: 67, 501: 247, 506: 54, 507: 126, 508: 179, 519: 119, 520: 104, 523: 21, 529: 285, 531: 131, 532: 10, 540: 228, 543: 186, 547: 213, 549: 231, 552: 14, 556: 222, 559: 111, 560: 252, 561: 18, 564: 156, 567: 55, 569: 40, 571: 240, 573: 196, 574: 206, 580: 103, 583: 126, 585: 100, 588: 222, 589: 19, 593: 151, 596: 88, 606: 83, 608: 27, 609: 193, 610: 78, 613: 29, 621: 23, 628: 179, 629: 186, 630: 119, 633: 252, 634: 235, 636: 23, 643: 63, 649: 246, 651: 101, 661: 142, 662: 109, 663: 222, 669: 256, 670: 19, 671: 230, 672: 109, 676: 265, 680: 43, 686: 96, 687: 133, 691: 222, 693: 78, 694: 203, 709: 256, 720: 130, 725: 101, 727: 150, 728: 129, 732: 9, 733: 28, 739: 130, 742: 92, 743: 118, 746: 249, 748: 221, 749: 244, 751: 236, 754: 233, 755: 70, 757: 273, 758: 213, 759: 20, 760: 224, 762: 272, 763: 233, 764: 123, 774: 101, 779: 135, 782: 137, 783: 127, 784: 122, 786: 286, 790: 216, 792: 157, 797: 278, 799: 151, 800: 250, 809: 84, 814: 70, 820: 247, 822: 194, 823: 294, 828: 280, 831: 115, 835: 20, 837: 284, 839: 253, 842: 143, 849: 115, 851: 137, 854: 291, 856: 179, 857: 34, 860: 45, 866: 89, 868: 249, 878: 297, 880: 201, 881: 29, 882: 145, 884: 24, 893: 31, 896: 42, 897: 91, 900: 190, 909: 292, 911: 161, 915: 279, 916: 158, 917: 57, 919: 244, 923: 184, 926: 247, 928: 201, 929: 18, 938: 86, 942: 52, 949: 46, 951: 149, 957: 47, 962: 90, 964: 250, 969: 231, 970: 296, 972: 250, 973: 77, 974: 154, 975: 5, 976: 73, 978: 234, 983: 232, 985: 297, 990: 220, 1008: 227, 1016: 193, 1017: 47, 1021: 185, 1022: 166, 1027: 134, 1030: 202, 1031: 252, 1032: 114, 1033: 249, 1035: 167, 1036: 52, 1041: 282, 1042: 36, 1044: 26, 1047: 275, 1050: 234, 1052: 253, 1053: 13, 1055: 232, 1056: 77, 1057: 216, 1061: 279, 1065: 110, 1067: 78, 1075: 63, 1081: 211, 1085: 255, 1086: 55, 1090: 45, 1091: 73, 1092: 8, 1094: 35, 1104: 279, 1107: 89, 1108: 201, 1109: 239, 1111: 10, 1113: 62, 1114: 149, 1115: 46, 1116: 204, 1122: 167, 1127: 126, 1129: 230, 1131: 289, 1135: 228, 1140: 139, 1142: 131, 1145: 277, 1155: 217, 1156: 222, 1163: 21, 1170: 53, 1173: 122, 1183: 190, 1186: 255, 1190: 127, 1191: 58, 1195: 108, 1199: 168, 1202: 199, 1205: 211, 1206: 56, 1208: 10, 1211: 149, 1219: 74, 1226: 48, 1227: 71, 1232: 183, 1234: 78, 1238: 200, 1239: 183, 1240: 52, 1243: 249, 1248: 101, 1251: 96, 1252: 117, 1259: 65, 1260: 124, 1261: 65, 1263: 119, 1264: 44, 1270: 251, 1274: 165, 1275: 291, 1276: 216, 1277: 80, 1279: 38, 1283: 32, 1284: 205, 1285: 277, 1286: 131, 1288: 13, 1290: 92, 1291: 23, 1296: 224, 1297: 127, 1302: 211, 1309: 141, 1310: 299, 1311: 29, 1312: 238, 1314: 94, 1316: 184, 1319: 142, 1320: 14, 1323: 64, 1325: 111, 1326: 42, 1327: 138, 1329: 137, 1331: 66, 1332: 27, 1340: 93, 1341: 178, 1342: 68, 1347: 60, 1359: 81, 1364: 234, 1368: 255, 1375: 185, 1378: 100, 1380: 150, 1381: 6, 1388: 261, 1389: 26, 1392: 228, 1397: 135, 1398: 117, 1400: 291, 1402: 24, 1404: 255, 1409: 184, 1410: 167, 1414: 18, 1415: 103, 1419: 23, 1420: 156, 1421: 264, 1423: 81, 1424: 141, 1435: 147, 1437: 185, 1440: 290, 1442: 240, 1443: 281, 1448: 90, 1449: 67, 1451: 297, 1452: 35, 1458: 143, 1463: 59, 1464: 78, 1465: 133, 1469: 116, 1471: 37, 1476: 236, 1478: 133, 1480: 93, 1482: 132, 1483: 178, 1485: 65, 1488: 87, 1490: 43, 1491: 127, 1493: 281, 1494: 286, 1496: 296, 1498: 237, 1499: 150, 1500: 275, 1504: 140, 1509: 6, 1510: 199, 1512: 139, 1514: 254, 1517: 180, 1520: 135, 1521: 54, 1523: 224, 1524: 24, 1525: 209, 1526: 95, 1528: 193, 1537: 36, 1545: 135, 1549: 142, 1555: 142, 1557: 103, 1559: 51, 1565: 283, 1571: 222, 1572: 76, 1575: 73, 1586: 159, 1588: 254, 1591: 92, 1600: 19, 1604: 55, 1606: 249, 1609: 96, 1619: 161, 1624: 81, 1629: 77, 1631: 158, 1633: 9, 1634: 225, 1640: 146, 1641: 239, 1645: 129, 1647: 285, 1648: 276, 1651: 230, 1653: 199, 1654: 297, 1658: 170, 1662: 279, 1665: 59, 1666: 18, 1670: 93, 1678: 10, 1679: 275, 1684: 14, 1686: 246, 1687: 205, 1689: 279, 1691: 72, 1695: 155, 1704: 232, 1705: 80, 1706: 227, 1716: 116, 1721: 239, 1722: 86, 1724: 80, 1725: 19, 1728: 269, 1730: 114, 1731: 204, 1733: 189, 1734: 35, 1735: 106, 1738: 289, 1740: 144, 1741: 290, 1745: 67, 1746: 283, 1748: 128, 1749: 273, 1765: 222, 1767: 124, 1771: 14, 1777: 116, 1778: 179, 1781: 225, 1782: 175, 1785: 128, 1790: 8, 1794: 141, 1799: 46, 1801: 222, 1807: 103, 1818: 201, 1825: 199, 1828: 53, 1832: 163, 1836: 265, 1838: 278, 1840: 100, 1845: 72, 1846: 132, 1847: 28, 1857: 190, 1860: 198, 1861: 202, 1865: 68, 1866: 255, 1867: 21, 1871: 91, 1873: 165, 1876: 23, 1878: 236, 1880: 16, 1882: 32, 1884: 254, 1888: 183, 1890: 286, 1894: 261, 1898: 253, 1899: 249, 1902: 173, 1906: 70, 1910: 145, 1913: 19, 1916: 93, 1918: 140, 1924: 39, 1925: 214, 1932: 150, 1936: 36, 1937: 235, 1940: 211, 1942: 107, 1947: 236, 1949: 247}}")]

In [19]:
df[df['Experience_Level'] == 'Expert'].groupby('Job_Completed')

In [30]:
# Фильтруем экспертов
df[(df['Experience_Level'] == 'Expert') & (df['Job_Completed'] < 100)].shape[0]
df[(df['Experience_Level'] == 'Expert')].shape[0]

641

In [29]:
round((
        df[(df['Experience_Level'] == 'Expert') & (df['Job_Completed'] < 100)].shape[0] / 
        df[(df['Experience_Level'] == 'Expert')].shape[0]
    ) * 100, 2)

33.85

In [67]:
experts = df[df['Experience_Level'] == 'Expert']
experts_distribution = (
    df[df['Experience_Level'] == 'Expert']
    .groupby('Job_Completed')
    .size()
    .reset_index(name='Количество_экспертов')
    .sort_values('Job_Completed')
)

# print("Распределение экспертов по количеству проектов:")
# print(experts_distribution)
# experts_distribution[['Job_Completed', 'Количество_экспертов']].to_dict()

# experts_distribution['Процент'] = round((experts_distribution['Количество_экспертов'] / len(experts) * 100), 2)
distribution_dict = dict(zip(experts_distribution['Job_Completed'], 
                            experts_distribution['Количество_экспертов']))
print(distribution_dict)

{5: 1, 6: 2, 8: 3, 9: 2, 10: 5, 13: 2, 14: 4, 16: 1, 17: 1, 18: 6, 19: 6, 20: 2, 21: 4, 23: 5, 24: 3, 25: 2, 26: 5, 27: 2, 28: 4, 29: 4, 30: 2, 31: 1, 32: 2, 33: 2, 34: 3, 35: 3, 36: 5, 37: 1, 38: 1, 39: 2, 40: 3, 41: 1, 42: 2, 43: 4, 44: 1, 45: 2, 46: 3, 47: 2, 48: 1, 51: 1, 52: 5, 53: 4, 54: 2, 55: 3, 56: 1, 57: 2, 58: 1, 59: 3, 60: 2, 62: 2, 63: 3, 64: 1, 65: 3, 66: 2, 67: 4, 68: 2, 70: 3, 71: 3, 72: 4, 73: 4, 74: 4, 75: 1, 76: 2, 77: 3, 78: 6, 80: 3, 81: 5, 83: 1, 84: 1, 86: 3, 87: 1, 88: 1, 89: 4, 90: 2, 91: 2, 92: 4, 93: 4, 94: 4, 95: 1, 96: 4, 99: 1, 100: 4, 101: 4, 102: 1, 103: 4, 104: 1, 106: 1, 107: 1, 108: 2, 109: 2, 110: 2, 111: 3, 112: 1, 114: 2, 115: 2, 116: 4, 117: 3, 118: 2, 119: 4, 120: 1, 122: 2, 123: 2, 124: 2, 125: 1, 126: 3, 127: 6, 128: 2, 129: 3, 130: 2, 131: 3, 132: 5, 133: 4, 134: 1, 135: 7, 137: 3, 138: 2, 139: 2, 140: 3, 141: 5, 142: 6, 143: 2, 144: 1, 145: 4, 146: 2, 147: 1, 149: 5, 150: 4, 151: 2, 153: 1, 154: 1, 155: 2, 156: 4, 157: 2, 158: 4, 159: 1, 160: